From 158b9cd571ab38cc9ae966df21e7242f04e1bd73 Mon Sep 17 00:00:00 2001
From: Stefano Stabellini <stefano.stabellini@amd.com>
Date: Sun, 11 Sep 2022 19:44:00 +0300
Subject: [PATCH 12/21] libxl: add support for emulated TPM on ARM

QEMU gained support for running as IOREQ server to do device emulation
for ARM guests. It is done using a "xenpv" emulated machine for aarch64
targets (same name as the x86 "xenpv" machine for PV backends.) It
supports both PV backends and IOREQs.

Introduce a new xl config option "tpm" to enable TPM emulation via QEMU.
We are not reusing the existing vtpm option because that is tied to a PV
frontend and backend protocol that has nothing to do with the TPM
emulation provided by QEMU. vTPM is configured via xenstore, while
emulated TPM via the QEMU command line.

If tpm=1 pass the right command line parameters to QEMU so that it can
connect to swtpm and provide TPM emulation for Xen guests. Also if tpm=1
request xenpv QEMU to be spawned.

Signed-off-by: Stefano Stabellini <stefano.stabellini@xilinx.com>
---
 docs/man/xl.cfg.5.pod.in         |  5 +++++
 tools/libs/light/libxl_arm.c     | 29 +++++++++++++++++++++++++++++
 tools/libs/light/libxl_dm.c      | 26 +++++++++++++++++++++++++-
 tools/libs/light/libxl_types.idl |  1 +
 tools/xl/xl_parse.c              |  1 +
 xen/include/public/arch-arm.h    |  4 ++++
 6 files changed, 65 insertions(+), 1 deletion(-)

diff --git a/docs/man/xl.cfg.5.pod.in b/docs/man/xl.cfg.5.pod.in
index 27c455210b..2ee59fe5d3 100644
--- a/docs/man/xl.cfg.5.pod.in
+++ b/docs/man/xl.cfg.5.pod.in
@@ -798,6 +798,11 @@ value is optional if this is a guest domain.
 
 =back
 
+=item B<tpm=BOOLEAN>
+
+B<(ARM only)> Specifies whether an B<emulated> TPM is enabled for the
+VM.
+
 =item B<p9=[ "9PFS_SPEC_STRING", "9PFS_SPEC_STRING", ...]>
 
 Creates a Xen 9pfs connection to share a filesystem from the backend to the
diff --git a/tools/libs/light/libxl_arm.c b/tools/libs/light/libxl_arm.c
index c1692e630c..c8b1bc85d6 100644
--- a/tools/libs/light/libxl_arm.c
+++ b/tools/libs/light/libxl_arm.c
@@ -961,6 +961,31 @@ static int make_vpci_node(libxl__gc *gc, void *fdt,
     return 0;
 }
 
+static int make_tpm_node(libxl__gc *gc, void *fdt,
+                         const struct arch_info *ainfo,
+                         struct xc_dom_image *dom)
+{
+    int res;
+
+    res = fdt_begin_node(fdt, "tpm_tis");
+    if (res) return res;
+
+    res = fdt_property_compat(gc, fdt, 1, "tcg,tpm-tis-mmio");
+    if (res) return res;
+
+    res = fdt_property_regs(gc, fdt, GUEST_ROOT_ADDRESS_CELLS,
+            GUEST_ROOT_SIZE_CELLS, 1, GUEST_TPM_BASE, GUEST_TPM_SIZE);
+    if (res) return res;
+
+    res = fdt_property_string(fdt, "status", "okay");
+    if (res) return res;
+
+    res = fdt_end_node(fdt);
+    if (res) return res;
+
+    return 0;
+}
+
 static int make_xen_iommu_node(libxl__gc *gc, void *fdt)
 {
     int res;
@@ -1464,6 +1489,9 @@ next_resize:
                                               libxl_defbool_val(virtio->grant_usage)) );
         }
 
+        if (libxl_defbool_val(d_config->b_info.tpm))
+            FDT( make_tpm_node(gc, fdt, ainfo, dom) );
+
         /*
          * The iommu node should be created only once for all virtio-mmio
          * devices.
@@ -1837,6 +1865,7 @@ int libxl__arch_domain_build_info_setdefault(libxl__gc *gc,
 {
     /* ACPI is disabled by default */
     libxl_defbool_setdefault(&b_info->acpi, false);
+    libxl_defbool_setdefault(&b_info->tpm, false);
 
     /* Trapping of unmapped accesses enabled by default.  */
     libxl_defbool_setdefault(&b_info->trap_unmapped_accesses, true);
diff --git a/tools/libs/light/libxl_dm.c b/tools/libs/light/libxl_dm.c
index 511ec76a65..09dfe3a0e3 100644
--- a/tools/libs/light/libxl_dm.c
+++ b/tools/libs/light/libxl_dm.c
@@ -1553,6 +1553,21 @@ static int libxl__build_device_model_args_new(libxl__gc *gc,
     for (i = 0; b_info->extra && b_info->extra[i] != NULL; i++)
         flexarray_append(dm_args, b_info->extra[i]);
 
+    /*
+     * swtpm needs to be started at system boot with the following
+     * command line parameters:
+     * swtpm socket --tpmstate dir=/tmp/vtpm2 --ctrl type=unixio,path=/tmp/vtpm2/swtpm-sock
+     *
+     * The socket path needs to correspond to the one passed to QEMU
+     * here.
+     */
+    if (libxl_defbool_val(b_info->tpm)) {
+        flexarray_append(dm_args, "-chardev");
+        flexarray_append(dm_args, "socket,id=chrtpm,path=/tmp/vtpm2/swtpm-sock");
+        flexarray_append(dm_args, "-tpmdev");
+        flexarray_append(dm_args, "emulator,id=tpm0,chardev=chrtpm");
+    }
+
     flexarray_append(dm_args, "-machine");
     switch (b_info->type) {
     case LIBXL_DOMAIN_TYPE_PVH:
@@ -3726,7 +3741,16 @@ int libxl__need_xenpv_qemu(libxl__gc *gc, libxl_domain_config *d_config)
         goto out;
     }
 
-    if (d_config->num_vfbs > 0) {
+    /*
+     * On ARM, the xenpv machine can be used for both PV backends and
+     * emulation. The only device emulation supported today is TPM
+     * emulation (in conjunction with SWTPM).
+     *
+     * Here also check for emulated TPM: if enabled xenpv QEMU is
+     * needed.
+     */
+    if (d_config->num_vfbs > 0 ||
+        libxl_defbool_val(d_config->b_info.tpm)) {
         ret = 1;
         goto out;
     }
diff --git a/tools/libs/light/libxl_types.idl b/tools/libs/light/libxl_types.idl
index ef49b3f98f..059bc0c8a0 100644
--- a/tools/libs/light/libxl_types.idl
+++ b/tools/libs/light/libxl_types.idl
@@ -650,6 +650,7 @@ libxl_domain_build_info = Struct("domain_build_info",[
     ("apic",             libxl_defbool),
     ("dm_restrict",      libxl_defbool),
     ("tee",              libxl_tee_type),
+    ("tpm",              libxl_defbool),
     ("u", KeyedUnion(None, libxl_domain_type, "type",
                 [("hvm", Struct(None, [("firmware",         string),
                                        ("bios",             libxl_bios_type),
diff --git a/tools/xl/xl_parse.c b/tools/xl/xl_parse.c
index 2cf8d7083e..85dd72845f 100644
--- a/tools/xl/xl_parse.c
+++ b/tools/xl/xl_parse.c
@@ -2866,6 +2866,7 @@ skip_usbdev:
     }
 
     xlu_cfg_get_defbool(config, "dm_restrict", &b_info->dm_restrict, 0);
+    xlu_cfg_get_defbool(config, "tpm", &b_info->tpm, 0);
 
     if (c_info->type == LIBXL_DOMAIN_TYPE_HVM) {
         if (!xlu_cfg_get_string (config, "vga", &buf, 0)) {
diff --git a/xen/include/public/arch-arm.h b/xen/include/public/arch-arm.h
index 095b1a23e3..1d675c55b9 100644
--- a/xen/include/public/arch-arm.h
+++ b/xen/include/public/arch-arm.h
@@ -471,6 +471,10 @@ typedef uint64_t xen_callback_t;
 #define GUEST_VPCI_MEM_ADDR                 xen_mk_ullong(0x23000000)
 #define GUEST_VPCI_MEM_SIZE                 xen_mk_ullong(0x10000000)
 
+/* Guest TPM device */
+#define GUEST_TPM_BASE    xen_mk_ullong(0x0c000000)
+#define GUEST_TPM_SIZE    xen_mk_ullong(0x00005000)
+
 /*
  * 16MB == 4096 pages reserved for guest to use as a region to map its
  * grant table in.
-- 
2.34.1

